### PR TITLE
perf: implement size hints for `Rect` iterators

### PR DESCRIPTION
**Pull request by airblast-dev**
Implement `Iter::size_hint` for `Rows`, `Columns`, and `Positions`.

The `size_hint` also accounts for the `DoubleEndedIterator` implementation, and provides an accurate hint when using `Iter`, and `DoubleEndedIterator`.

Performance wise this doesn't change iteration in any way. However it optimizes collecting the items in a collection (`Vec`, `HashSet`, `HashMap`). 

<details>

<summary>Performance can be measured  by applying following patch, and then running "cargo bench rect":</summary>

```patch
diff --git a/benches/main/rect.rs b/benches/main/rect.rs
index 828bfc3..f273454 100644
--- a/benches/main/rect.rs
+++ b/benches/main/rect.rs
@@ -1,24 +1,113 @@
-use criterion::{black_box, criterion_group, BenchmarkId, Criterion};
+use criterion::{black_box, criterion_group, BatchSize, Bencher, BenchmarkId, Criterion};
 use ratatui::layout::Rect;

-fn rect_rows_benchmark(c: &mut Criterion) {
-    let rect_sizes = vec![
-        Rect::new(0, 0, 1, 16),
-        Rect::new(0, 0, 1, 1024),
-        Rect::new(0, 0, 1, 65535),
-    ];
-    let mut group = c.benchmark_group("rect_rows");
-    for rect in rect_sizes {
-        group.bench_with_input(BenchmarkId::new("rows", rect.height), &rect, |b, rect| {
-            b.iter(|| {
-                for row in rect.rows() {
-                    // Perform any necessary operations on each row
-                    black_box(row);
-                }
-            });
-        });
+fn rect_iters_benchmark(c: &mut Criterion) {
+    let rect_sizes = vec![[16, 16], [64, 64], [255, 255]];
+    let mut group = c.benchmark_group("rect");
+    // We don't use `Rect::new` as it will clamp the value to a max of 255x255.
+    for rect in rect_sizes.into_iter().map(|[width, height]| Rect {
+        width,
+        height,
+        ..Default::default()
+    }) {
+        group.bench_with_input(
+            BenchmarkId::new("rect_rows_iter", rect.height),
+            &rect,
+            |b, rect| rect_rows_iter(b, *rect),
+        );
+        group.bench_with_input(
+            BenchmarkId::new("rect_rows_collect", rect.height),
+            &rect,
+            |b, rect| rect_rows_collect(b, *rect),
+        );
+        group.bench_with_input(
+            BenchmarkId::new("rect_columns_iter", rect.width),
+            &rect,
+            |b, rect| rect_columns_iter(b, *rect),
+        );
+        group.bench_with_input(
+            BenchmarkId::new("rect_columns_collect", rect.width),
+            &rect,
+            |b, rect| rect_columns_collect(b, *rect),
+        );
+        group.bench_with_input(
+            BenchmarkId::new(
+                "rect_positions_iter",
+                format!("{}x{}", rect.width, rect.height),
+            ),
+            &rect,
+            |b, rect| rect_positions_iter(b, *rect),
+        );
+        group.bench_with_input(
+            BenchmarkId::new(
+                "rect_positions_collect",
+                format!("{}x{}", rect.width, rect.height),
+            ),
+            &rect,
+            |b, rect| rect_positions_collect(b, *rect),
+        );
     }
     group.finish();
 }

-criterion_group!(benches, rect_rows_benchmark);
+fn rect_rows_iter(c: &mut Bencher, rect: Rect) {
+    c.iter_batched(
+        || black_box(rect),
+        |rect| {
+            for row in black_box(rect.rows()) {
+                black_box(row);
+            }
+        },
+        BatchSize::LargeInput,
+    );
+}
+
+fn rect_rows_collect(c: &mut Bencher, rect: Rect) {
+    c.iter_batched(
+        || black_box(rect),
+        |rect| black_box(rect.rows()).collect::<Vec<_>>(),
+        BatchSize::LargeInput,
+    );
+}
+
+fn rect_columns_iter(c: &mut Bencher, rect: Rect) {
+    c.iter_batched(
+        || black_box(rect),
+        |rect| {
+            for col in black_box(rect.columns()) {
+                black_box(col);
+            }
+        },
+        BatchSize::LargeInput,
+    );
+}
+
+fn rect_columns_collect(c: &mut Bencher, rect: Rect) {
+    c.iter_batched(
+        || black_box(rect),
+        |rect| black_box(rect.columns()).collect::<Vec<_>>(),
+        BatchSize::LargeInput,
+    );
+}
+
+fn rect_positions_iter(c: &mut Bencher, rect: Rect) {
+    c.iter_batched(
+        || black_box(rect),
+        |rect| {
+            for pos in black_box(rect.positions()) {
+                black_box(pos);
+            }
+        },
+        BatchSize::LargeInput,
+    );
+}
+
+fn rect_positions_collect(b: &mut Bencher, rect: Rect) {
+    b.iter_batched(
+        || black_box(rect),
+        |rect| black_box(rect.positions()).collect::<Vec<_>>(),
+        BatchSize::LargeInput,
+    );
+}
+
+criterion_group!(benches, rect_iters_benchmark);
```
</details>

The exact performance in actual usage can vary. This change improves the performance of collecting the items by around 50-70%. With a u16::MAX provided to any of the iterators this drops substantially. This only drastically improved performance for lower inputs such as 1024 items. 

<details>

<summary>Benchmark result compared to without size hint</summary>

```
rect/rect_rows_iter/16  time:   [5.3884 ns 5.3892 ns 5.3904 ns]
                        change: [+0.0051% +0.1901% +0.3621%] (p = 0.03 < 0.05)
                        Change within noise threshold.
Found 20 outliers among 100 measurements (20.00%)
  3 (3.00%) high mild
  17 (17.00%) high severe
rect/rect_rows_collect/16
                        time:   [24.130 ns 24.162 ns 24.191 ns]
                        change: [-49.206% -48.831% -48.432%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  6 (6.00%) high mild
  9 (9.00%) high severe
rect/rect_columns_iter/16
                        time:   [6.8879 ns 6.8956 ns 6.9038 ns]
                        change: [-1.2312% -1.0829% -0.9368%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
rect/rect_columns_collect/16
                        time:   [25.824 ns 25.849 ns 25.879 ns]
                        change: [-50.596% -50.254% -49.914%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe
Benchmarking rect/rect_positions_iter/16x16: Collecting 100 samples in estimated 5.0003 s (29M iterationsrect/rect_positions_iter/16x16
                        time:   [172.08 ns 172.09 ns 172.10 ns]
                        change: [-0.2681% -0.1776% -0.0955%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) high mild
  7 (7.00%) high severe
Benchmarking rect/rect_positions_collect/16x16: Collecting 100 samples in estimated 5.0009 s (24M iteratirect/rect_positions_collect/16x16
                        time:   [200.75 ns 200.95 ns 201.23 ns]
                        change: [-46.206% -45.987% -45.763%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  3 (3.00%) high severe
rect/rect_rows_iter/64  time:   [14.952 ns 14.959 ns 14.964 ns]
                        change: [-0.5365% -0.1234% +0.1449%] (p = 0.61 > 0.05)
                        No change in performance detected.
rect/rect_rows_collect/64
                        time:   [55.562 ns 55.615 ns 55.676 ns]
                        change: [-58.737% -58.527% -58.228%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe
rect/rect_columns_iter/64
                        time:   [17.975 ns 17.993 ns 18.008 ns]
                        change: [-1.0430% -0.7802% -0.4111%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
rect/rect_columns_collect/64
                        time:   [60.467 ns 60.506 ns 60.554 ns]
                        change: [-61.007% -60.821% -60.621%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe
Benchmarking rect/rect_positions_iter/64x64: Collecting 100 samples in estimated 5.0084 s (1.9M iterationrect/rect_positions_iter/64x64
                        time:   [2.6433 µs 2.6439 µs 2.6447 µs]
                        change: [-0.7008% -0.2778% -0.0235%] (p = 0.10 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe
Benchmarking rect/rect_positions_collect/64x64: Collecting 100 samples in estimated 5.0062 s (1.9M iteratrect/rect_positions_collect/64x64
                        time:   [2.6869 µs 2.6874 µs 2.6879 µs]
                        change: [-15.808% -15.498% -15.270%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 19 outliers among 100 measurements (19.00%)
  8 (8.00%) low severe
  5 (5.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe
rect/rect_rows_iter/255 time:   [55.839 ns 55.857 ns 55.879 ns]
                        change: [-0.0632% +0.0407% +0.1614%] (p = 0.54 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe
rect/rect_rows_collect/255
                        time:   [173.84 ns 173.99 ns 174.18 ns]
                        change: [-40.416% -40.260% -40.113%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
rect/rect_columns_iter/255
                        time:   [58.862 ns 58.895 ns 58.949 ns]
                        change: [+0.8115% +0.9726% +1.1506%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe
rect/rect_columns_collect/255
                        time:   [199.95 ns 200.09 ns 200.25 ns]
                        change: [-40.872% -40.593% -40.327%] (p = 0.00 < 0.05)
                        Performance has improved.
Benchmarking rect/rect_positions_iter/255x255: Collecting 100 samples in estimated 5.0715 s (121k iteratirect/rect_positions_iter/255x255
                        time:   [41.845 µs 41.846 µs 41.847 µs]
                        change: [-0.4778% -0.4015% -0.3222%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe
Benchmarking rect/rect_positions_collect/255x255: Collecting 100 samples in estimated 5.1011 s (121k iterrect/rect_positions_collect/255x255
                        time:   [42.034 µs 42.046 µs 42.061 µs]
                        change: [-7.5948% -7.4331% -7.2038%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) high mild
  5 (5.00%) high severe
```
</details>

### CODE REVIEW

**Comment by joshka** [comment]
Feel free to add benchmarks, but as a general rule, only worry about using multiple sizes if the absolute magnitude of the different sizes is on a meaningful human relevant scale (e.g. some amount of milliseconds would be relevant, some amount of nanoseconds would not). For rect related stuff, I'd choose maybe two values 200x50 (a neat terminal size) and u16::Max (this is likely not relevant to normal use cases, but it covers the worst case).

The easiest approach to benchmark comparisons is to put the bench in a single commit stacked before the changes (use interactive rebase is needed to sort this out / use fixup commits and autosquash to update). Then run the bench before then checkout the last commit and run again. Criterion gives nice deltas.

File: src/layout/rect/iter.rs [hunk 1]
```diff
@@ -35,6 +35,19 @@ impl Iterator for Rows {
...
+        let remaining_count = self
+            .rect
+            .height
+            // Subtract the number of rows that were provided from the start of the `Rect`, from the
+            // total height.
+            .saturating_sub(self.current_row_fwd.saturating_sub(self.rect.y))
+            // Subtract the number of rows that were provided from the end of the `Rect`, from the
+            // total height.
+            .saturating_sub(self.rect.bottom().saturating_sub(self.current_row_back));
+        (remaining_count as usize, None)
```
Comment (joshka) [inline src/layout/rect/iter.rs:49]: nit: variable names communicate the same thing as the comments pretty well here (and below)
Should the second element of the return value be `Some(count)` here? I think I'm reading https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint right.

```suggestion
        let start_count = self.current_row_fwd.saturating_sub(self.rect.top());
        let end_count = self.rect.bottom().saturating_sub(self.current_row_back);
        let count = self
            .rect
            .height
            .saturating_sub(start_count)
            .saturating_sub(end_count) as usize;
        (count, Some(count))
```

**Review by joshka — APPROVED** [review]
Generally LGTM. Minor nits only and wouldn't mind seeing the benchmark added as mentioned in the main part of this PR.

File: src/layout/rect/iter.rs (hunk 1 above)
Comment (airblast-dev) [reply src/layout/rect/iter.rs:49]: You are correct, on returning `Some(count)`.  Will clear up the comments as well.

**Comment by airblast-dev** [comment]
I specifically removed the `u16::MAX` from the benchmark as part of the patch. This was due to `Rect::new` clamping, and `Rect::positions` taking pretty long on my machine when provided a `u16::MAX`. We can lower the sample count, or use a lower value than `u16::MAX`, but I am more inclined to use realistic values instead.

As for the benchmark and results, I will update my initial comment in a bit.

**Review by joshka — APPROVED** [review]
I simplified the functions a bit in 397822aa5 for readability.

Looking at the absolute values of the benchmark, I'm not too concerned about the perf gain on these as the worst bench was 42µs with non-representative data. Worth having, but not worth worrying too much more about. Maybe in a large virtual buffer for scrolling, or something like that this would cause measurable pain, but that's about it. Keeping open for the benchmarks to be added

**Comment by airblast-dev** [comment]
> ...42µs with non-representative data...
I definitely agree that the performance isn't anything to write home about. Though for relatively simple iterators I think size_hint is a reasonable thing to implement.

Will add the benchmarks shortly.

**Review by orhun — APPROVED** [review]
+1 from me, we can probably merge after the benchmarks are added :)

**Review by joshka — APPROVED** [review]

**Comment by joshka** [comment]
Thanks again for the PR

**Pull request merged**